numpy/ma/core.py

```diff
@@ -4286,22 +4297,45 @@ def count(self, axis=None):
         array([3, 0])
 
         """
+        kwargs = {} if keepdims is np._NoValue else {'keepdims': keepdims}
+
         m = self._mask
-        s = self.shape
+        # special case for matrices (we assume no other subclasses modify
+        # their dimensions)
+        if isinstance(self.data, np.matrix):
+            if m is nomask:
+                m = np.zeros(self.shape, dtype=np.bool_)
+            m = m.view(type(self.data))
+
         if m is nomask:
-            if axis is None:
+            # compare to _count_reduce_items in _methods.py
+
+            if self.shape is ():
+                return 1
```

Comment:
Oh sorry I was inaccurate in my last comment. Here's a better description of the old behavior:

 * `count()` gives no error
 * `count(axis=0)` and `count(axis=1)` both give IndexError

In the new behavior
 * `count()` and `count(axis=0)` give no error
 * `count(axis=1)` gives a ValueError

I think the old and new behavior are similar enough, and this is such a small edge case, that I'm ok with just making the change.